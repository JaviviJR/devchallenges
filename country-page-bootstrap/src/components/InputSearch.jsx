import PropTypes from 'prop-types';
import styles from './InputSearch.module.css';
import Search from "../assets/Search.svg"
import { useEffect, useState } from 'react';

function InputSearch({ handler }) {
    const [search, setSearch] = useState('');

    useEffect(() => {
        const timeout = setTimeout(() => {
            handler(search);
        }, 500);
        return () => clearTimeout(timeout);
    }, [search]);

    return (
        <div className={`input-group ${styles.inputSearch}`}>
            <span className='input-group-text'><img src={ Search } /></span>
            <input 
                type='text'
                className='form-control'
                placeholder='Search by Name, Region, Subregion'
                onChange={ (event) => setSearch(event.target.value)}
            />
        </div>
    );
}

export default InputSearch;

InputSearch.propTypes = {
    handler: PropTypes.func.isRequired
}