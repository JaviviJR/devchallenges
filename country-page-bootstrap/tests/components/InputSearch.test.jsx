import { render, screen, waitFor } from "@testing-library/react"
import userEvent from '@testing-library/user-event'
import InputSearch from "../../src/components/InputSearch";

jest.mock('../../src/assets/Search.svg', () => 'Search icon');

describe('testing InputSearch component', () => {
    const handlerMock = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should match with snapshot', () => {
        const { container } = render(
            <InputSearch handler={ handlerMock } />
        );

        expect(container).toMatchSnapshot();
    })

    test('should show input text', () => {
        render(
            <InputSearch handler={ handlerMock } />
        );
        
        const input = screen.getByRole('textbox');
        expect( input ).toBeInTheDocument();
    });

    test('should handle search text', async () => {
        render(
            <InputSearch handler={ handlerMock } />
        );

        const input = screen.getByRole('textbox');
        
        await userEvent.type(input, 't');
        await new Promise((resolve) => setTimeout(resolve, 50));
        await userEvent.type(input, 'e');
        await new Promise((resolve) => setTimeout(resolve, 50));
        await userEvent.type(input, 's');
        await new Promise((resolve) => setTimeout(resolve, 50));
        await userEvent.type(input, 't');

        // Wait for the debounce timeout to finish
        await waitFor(() => expect(handlerMock).toHaveBeenCalledTimes(1));
        expect(handlerMock).toHaveBeenCalledWith('test');
    });
})