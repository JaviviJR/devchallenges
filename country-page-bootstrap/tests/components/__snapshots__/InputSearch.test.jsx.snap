// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing InputSearch component should match with snapshot 1`] = `
<div>
  <div
    class="input-group inputSearch"
  >
    <span
      class="input-group-text"
    >
      <img
        src="Search icon"
      />
    </span>
    <input
      class="form-control"
      placeholder="Search by Name, Region, Subregion"
      type="text"
    />
  </div>
</div>
`;
